File: aec-project/core/delay_estimator.py
"""
delay_estimator.py - Uoc luong do tre va can chinh tin hieu reference

Module nay giai quyet bai toan: tin hieu reference x(n) (lay tu buffer loa)
va echo trong mic d(n) co do tre D mau khong biet truoc, do:
  - Buffer latency cua driver am thanh (~20-50ms)
  - Thoi gian truyen am tu loa den mic (khoang 1ms / 34cm)

Neu khong uoc luong va bu D, NLMS se khong hoi tu duoc vi cua so bo loc
khong khop voi vi tri echo thuc te trong tin hieu mic.

Gom 2 class:
  1. DelayEstimator: Dung GCC-PHAT de tinh D tu tuong quan cheo.
  2. DelayLine: Ring buffer de dich cham tin hieu reference di D mau.

GCC-PHAT (Generalized Cross-Correlation with Phase Transform):
    R(f) = X(f) * conj(D(f))
    R_phat(f) = R(f) / |R(f)|        <-- chi giu thong tin pha, bo bien do
    gcc(tau) = IFFT(R_phat)
    delay = argmax(gcc(tau))

    PHAT weighting chi giu pha nen peak rat nhon, chinh xac hon
    cross-correlation thong thuong (khong bi anh huong boi bien do tin hieu).

Ky thuat toi uu ap dung:
  - Tich luy nhieu frame (acc_frames=4) truoc khi tinh GCC -> giam jitter
  - Lam min pho GCC-PHAT (smooth_alpha=0.9) -> delay on dinh sau hoi tu
  - Zero-padding (2*N) -> tranh nhieu circular correlation
"""

import numpy as np


class DelayEstimator:
    """Uoc luong do tre bang GCC-PHAT.

    Tich luy acc_frames frame truoc khi tinh, sau do lam min pho GCC-PHAT
    theo thoi gian. Ket qua la delay on dinh, khong bi jitter.
    """

    def __init__(
        self,
        sample_rate: int = 16000,
        max_delay_ms: float = 250.0,
        smooth_alpha: float = 0.9,
        acc_frames: int = 4,
    ) -> None:
        self.sr = sample_rate

        # Gioi han tim kiem: chi xet delay trong khoang [0, max_delay_samples].
        # 250ms o 16kHz = 4000 mau. Du cho hau het he thong desktop.
        self.max_delay_samples = int(max_delay_ms * sample_rate / 1000)

        # He so lam min pho GCC-PHAT. Gia tri cao (0.9) -> on dinh nhung
        # phan ung cham voi thay doi. Gia tri thap -> nhay nhung nhieu jitter.
        self.alpha = smooth_alpha

        # So frame tich luy truoc khi thuc hien GCC-PHAT.
        # Tich luy nhieu frame -> tin hieu dai hon -> peak GCC ro rang hon.
        self.acc_frames = acc_frames

        # Buffer tich luy tin hieu reference va mic
        self._ref_acc: list[np.ndarray] = []
        self._mic_acc: list[np.ndarray] = []

        # Pho GCC-PHAT da lam min (duy tri qua cac lan goi)
        self._R_phat_smooth = None

        # Ket qua delay hien tai (so mau)
        self._current_delay = 0

    def update(self, ref_frame: np.ndarray, mic_frame: np.ndarray) -> int:
        """Cap nhat uoc luong delay voi cap frame moi.

        Tich luy du acc_frames frame roi moi tinh GCC-PHAT.
        Giua cac lan tinh, tra ve delay cu (amortized cost).

        Args:
            ref_frame: Tin hieu reference shape (N,)
            mic_frame: Tin hieu mic shape (N,)

        Returns:
            delay: So mau tre uoc luong (>= 0)
        """
        self._ref_acc.append(ref_frame.astype(np.float64))
        self._mic_acc.append(mic_frame.astype(np.float64))

        if len(self._ref_acc) >= self.acc_frames:
            # Noi tat ca frame da tich luy thanh tin hieu dai
            x = np.concatenate(self._ref_acc)
            d = np.concatenate(self._mic_acc)
            self._ref_acc.clear()
            self._mic_acc.clear()

            # Zero-padding (2*N): bat buoc de tranh nhieu do circular correlation.
            # Khi FFT khong co zero-pad, tuong quan vong (circular) co the tao
            # peak gia vi tin hieu "cuon vong" dau-cuoi.
            n_fft = 2 * len(x)
            X = np.fft.rfft(x, n=n_fft)
            D = np.fft.rfft(d, n=n_fft)

            # Tinh cross-spectrum va ap dung PHAT weighting (chi giu pha)
            R = X * np.conj(D)
            magnitude = np.abs(R)
            magnitude = np.maximum(magnitude, 1e-10)  # Tranh chia 0
            R_phat = R / magnitude

            # Lam min pho GCC-PHAT qua thoi gian.
            # Khac voi cach lam min gia tri delay (bi nhay lien tuc),
            # lam min truc tiep tren pho giup peak GCC dan sac net hon,
            # va delay hoi tu ve gia tri on dinh tuyet doi (khong jitter).
            if self._R_phat_smooth is None:
                self._R_phat_smooth = R_phat
            else:
                self._R_phat_smooth = (
                    self.alpha * self._R_phat_smooth
                    + (1.0 - self.alpha) * R_phat
                )

            # IFFT de chuyen ve mien thoi gian -> tim peak
            gcc = np.fft.irfft(self._R_phat_smooth, n=n_fft)

            # Chi tim trong khoang [0, max_delay_samples]
            search_range = gcc[:self.max_delay_samples + 1]
            self._current_delay = int(np.argmax(search_range))

        return self._current_delay

    @property
    def current_delay_samples(self) -> int:
        """Delay hien tai tinh bang so mau."""
        return self._current_delay

    @property
    def current_delay_ms(self) -> float:
        """Delay hien tai tinh bang mili-giay."""
        return self._current_delay * 1000.0 / self.sr


class DelayLine:
    """Ring buffer de dich cham (delay) tin hieu reference.

    Khi DelayEstimator tinh duoc delay = D mau, DelayLine se tra ve
    x(n - D) thay vi x(n). Cach lam: ghi mau moi vao buffer tai write_idx,
    doc mau cu tai (write_idx - D), ca hai quay vong trong mang co dinh.

    Uu diem so voi np.roll hay concatenate:
      - Khong cap phat bo nho moi moi frame
      - Delay co the thay doi bat ky luc nao ma khong mat du lieu
      - Hoat dong lien tuc qua cac frame (stateful)
    """

    def __init__(self, max_delay_samples: int = 48000) -> None:
        # max_delay = 48000 mau = 3 giay o 16kHz. Du cho moi kich ban thuc te.
        self.max_delay = max_delay_samples
        self.buffer = np.zeros(self.max_delay, dtype=np.float64)
        self.write_idx = 0

    def process(self, frame: np.ndarray, delay: int) -> np.ndarray:
        """Dich cham frame di delay mau.

        Args:
            frame: Tin hieu dau vao shape (N,)
            delay: So mau can tre (>=0)

        Returns:
            out: Tin hieu da duoc tre delay mau, shape (N,)
        """
        N = len(frame)
        out = np.zeros(N, dtype=np.float64)

        # Clamp delay de tranh truy cap ngoai vung buffer
        delay = int(np.clip(delay, 0, self.max_delay - 1))

        for i in range(N):
            # Ghi mau moi vao vi tri hien tai
            self.buffer[self.write_idx] = frame[i]
            # Doc mau cu cach delay buoc ve truoc (modulo de quay vong)
            read_idx = (self.write_idx - delay) % self.max_delay
            out[i] = self.buffer[read_idx]
            self.write_idx = (self.write_idx + 1) % self.max_delay

        return out

    def reset(self) -> None:
        """Xoa toan bo buffer. Goi khi bat dau phien moi."""
        self.buffer[:] = 0.0
        self.write_idx = 0

File: aec-project/mobile-app/ios/echocancellation/AEC/DelayEstimator.swift
import Foundation

final class DelayEstimator {
    private let sampleRate: Int
    private let maxDelaySamples: Int
    private let alpha: Float
    private let accFrames: Int
    private let minConfidence: Float
    private let confirmCount: Int
    private let nFFT: Int
    private let dft: DFT

    private var refAcc: [[Float]] = []
    private var micAcc: [[Float]] = []
    private var rPhatSmooth: (real: [Float], imag: [Float])?
    private var currentDelay: Int = 0

    private var candidateDelay: Int = 0
    private var candidateCount: Int = 0
    private var locked: Bool = false

    init(
        sampleRate: Int = 16_000,
        frameSize: Int = 1_024,
        maxDelayMs: Float = 250,
        smoothAlpha: Float = 0.9,
        accFrames: Int = 4,
        minConfidence: Float = 0.15,
        confirmCount: Int = 3
    ) {
        self.sampleRate = sampleRate
        self.maxDelaySamples = Int(maxDelayMs * Float(sampleRate) / 1000)
        self.alpha = smoothAlpha
        self.accFrames = accFrames
        self.minConfidence = minConfidence
        self.confirmCount = confirmCount
        self.nFFT = 2 * frameSize * accFrames
        self.dft = DFT(length: self.nFFT)!
    }

    func update(refFrame: [Float], micFrame: [Float]) -> Int {
        refAcc.append(refFrame)
        micAcc.append(micFrame)

        if refAcc.count >= accFrames {
            let x = refAcc.flatMap { $0 }
            let d = micAcc.flatMap { $0 }
            refAcc.removeAll(keepingCapacity: true)
            micAcc.removeAll(keepingCapacity: true)

            let refEnergy = DSPMath.meanPower(x)
            let micEnergy = DSPMath.meanPower(d)
            if refEnergy < 1e-7 || micEnergy < 1e-7 {
                return currentDelay
            }

            var xPad = x
            xPad.append(contentsOf: [Float](repeating: 0, count: max(0, nFFT - x.count)))
            var dPad = d
            dPad.append(contentsOf: [Float](repeating: 0, count: max(0, nFFT - d.count)))

            let xF = dft.forwardReal(xPad)
            let dF = dft.forwardReal(dPad)

            var rReal = [Float](repeating: 0, count: nFFT)
            var rImag = [Float](repeating: 0, count: nFFT)
            for i in 0..<nFFT {
                let xr = xF.real[i]
                let xi = xF.imag[i]
                let dr = dF.real[i]
                let di = dF.imag[i]
                // X * conj(D)
                let real = xr * dr + xi * di
                let imag = xi * dr - xr * di
                let mag = max(1e-10, hypot(real, imag))
                rReal[i] = real / mag
                rImag[i] = imag / mag
            }

            if var smooth = rPhatSmooth {
                for i in 0..<nFFT {
                    smooth.real[i] = alpha * smooth.real[i] + (1 - alpha) * rReal[i]
                    smooth.imag[i] = alpha * smooth.imag[i] + (1 - alpha) * rImag[i]
                }
                rPhatSmooth = smooth
            } else {
                rPhatSmooth = (rReal, rImag)
            }

            let smooth = rPhatSmooth!
            let gcc = dft.inverseComplex(real: smooth.real, imag: smooth.imag)
            let searchEnd = min(maxDelaySamples, gcc.count - 1)
            if searchEnd <= 0 { return currentDelay }
            let search = Array(gcc[0...searchEnd])
            guard let peak = search.enumerated().max(by: { $0.element < $1.element }) else {
                return currentDelay
            }
            let rawDelay = peak.offset
            let peakVal = peak.element
            let meanVal = max(1e-10, search.map { abs($0) }.reduce(0, +) / Float(search.count))
            let confidence = peakVal / meanVal

            if confidence < minConfidence {
                return currentDelay
            }

            if !locked {
                if abs(rawDelay - candidateDelay) <= 2 {
                    candidateCount += 1
                } else {
                    candidateDelay = rawDelay
                    candidateCount = 1
                }
                if candidateCount >= confirmCount {
                    currentDelay = candidateDelay
                    locked = true
                }
            } else {
                let atLocked = search[min(currentDelay, search.count - 1)]
                let lockStillGood = atLocked > 0.5 * peakVal
                if !lockStillGood {
                    if abs(rawDelay - candidateDelay) <= 2 {
                        candidateCount += 1
                    } else {
                        candidateDelay = rawDelay
                        candidateCount = 1
                    }
                    if candidateCount >= confirmCount {
                        currentDelay = candidateDelay
                        candidateCount = 0
                    }
                }
            }
        }

        return currentDelay
    }

    var currentDelaySamples: Int { currentDelay }
    var currentDelayMs: Float { Float(currentDelay) * 1000 / Float(sampleRate) }
}

final class DelayLine {
    private let maxDelay: Int
    private var buffer: [Float]
    private var writeIndex: Int = 0

    init(maxDelaySamples: Int = 48_000) {
        self.maxDelay = maxDelaySamples
        self.buffer = [Float](repeating: 0, count: maxDelaySamples)
    }

    func process(_ frame: [Float], delay: Int) -> [Float] {
        let n = frame.count
        let d = max(0, min(maxDelay - 1, delay))
        var out = [Float](repeating: 0, count: n)
        for i in 0..<n {
            buffer[writeIndex] = frame[i]
            let readIndex = (writeIndex - d + maxDelay) % maxDelay
            out[i] = buffer[readIndex]
            writeIndex = (writeIndex + 1) % maxDelay
        }
        return out
    }

    func reset() {
        buffer = [Float](repeating: 0, count: maxDelay)
        writeIndex = 0
    }
}
